### PR TITLE
Bayesian Gaussian Mixture Models, using variational inference

### PR DESCRIPTION
**Pull request by mamikonyana**
@pbailis probably it's better to discuss this first

### CODE REVIEW

File: src/main/java/macrobase/analysis/stats/VariationalGMM.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        priorW = MatrixUtils.createRealIdentityMatrix(dimensions);
+        RealMatrix priorWInverse = AlgebraUtils.createInverseMatrix(priorW);
+
+        alpha = new double[K];
+        beta = new double[K];
+        nu = new double[K];
+        W = new ArrayList<>(K);
+
+        // Initialize
+        m = gonzalezInitilizeMixtureCenters(data, K);
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/VariationalGMM.java]: typo: Initialize

File: src/main/java/macrobase/analysis/stats/VariationalGMM.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                    r[n][k] /= normalizingConstant;
+                    // Calculate unnormalized cluster weight, cluster mean
+                    clusterWeight[k] += r[n][k];
+                    clusterMean.set(k, clusterMean.get(k).add(data.get(n).getMetrics().mapMultiply(r[n][k])));
+                }
+            }
+
+
+            // 2. Reevaluate clusters based on densities that we have for each point.
+            for (int k=0; k<this.K; k++) {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/VariationalGMM.java]: minor style nit -- spacing...

File: src/main/java/macrobase/analysis/stats/VariationalGMM.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                }
+                S.set(k, S.get(k).scalarMultiply(1 / clusterWeight[k]));
+            }
+
+            for (int k=0; k<this.K; k++) {
+                alpha[k] = priorAlpha + clusterWeight[k];
+                beta[k] = priorBeta + clusterWeight[k];
+                m.set(k, priorM.mapMultiply(priorBeta).add(clusterMean.get(k).mapMultiply(clusterWeight[k])).mapDivide(beta[k]));
+                nu[k] = priorNu + 1 + clusterWeight[k];
+                RealVector adjastedMean = clusterMean.get(k).subtract(priorM);
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/VariationalGMM.java]: adjustedMean

File: src/main/java/macrobase/analysis/stats/VariationalGMM.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            log.debug("cluster means are at {}", clusterMean);
+            log.debug("cluster weights are at {}", clusterWeight);
+
+            double oldLogLikelihood = logLikelihood;
+            logLikelihood = 0;
+            for (int n = 0; n < N; n++) {
+                logLikelihood += Math.log(score(data.get(n)));
+            }
+
+            log.info("log likelihood after iteration {} is {}", iteration, logLikelihood);
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/VariationalGMM.java]: do you want this as info or trace?

File: src/main/java/macrobase/analysis/stats/VariationalGMM.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+            double improvement = (logLikelihood - oldLogLikelihood) / ( -logLikelihood);
+            if (improvement >= 0 && improvement < this.progressCutoff) {
+                log.debug("Breaking because imporovemnt was {} percent", improvement * 100);
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/VariationalGMM.java]: typo

**Comment by pbailis** [comment]
Two high level comments:

1. There are a few typos and weird formatting issues. Can you run `Reformat Code` in IntelliJ?
2. As we discussed on the phone, can we refactor all of these bayesian methods into their own subpackage?

**Comment by mamikonyana** [comment]
I didn't know about `Reformat Code`, it's really nice feature.
I'll move things into a sub package in the next pull request, I already have done it in a different branch and don't want commit merges.

**Pull request closed without merging**